drift/core/psmc.py

```diff
@@ -209,6 +213,273 @@ def _work_fisher_bias_m(self, mi):
+
+    def generate(self, regen=False):
+        """Calculate the total Fisher matrix and bias and save to a file.
+
+        Parameters
+        ----------
+        regen : boolean, optional
+            Force regeneration if products already exist (default `False`).
+        """
+        if mpiutil.rank0:
+            st = time.time()
+            print("======== Starting PS calculation MC XLARGE ========")
+
+        ffile = self.psdir + "/fisher.hdf5"
+
+        if os.path.exists(ffile) and not regen:
+            print("Fisher matrix file: %s exists. Skipping..." % ffile)
+            return
+
+        mpiutil.barrier()
+
+        # Pre-compute all the angular power spectra for the bands.
+        self.genbands()
+
+        comm = MPI.COMM_WORLD
+        size = comm.Get_size()
+        rank = comm.Get_rank()
+
+        st = time.time()
+
+        m_chunks, low_bound, upp_bound = self.split_single(
+            (self.telescope.mmax + 1), size
+        )
+        num_chunks = m_chunks.shape[0]
+
+        if mpiutil.rank0:
+            print(
+                "M chunks, lower bounds, upper bounds", m_chunks, low_bound, upp_bound
+            )
+
+        n = self.nsamples
+
+        self.qa = mpiarray.MPIArray(
+            (self.telescope.mmax + 1, self.nbands, n),
+            axis=1,
+            comm=MPI.COMM_WORLD,
+            dtype=np.float64,
+        )
+
+        self.qa[:] = 0.0
+
+        # This is designed such that at each iteration one rank gets one m-mode.
+        for ci, num_m in enumerate(m_chunks):
+            if mpiutil.rank0:
+                print("Starting chunk %i of %i" % (ci, num_chunks))
+
+            loc_num, loc_start, loc_end = mpiutil.split_local(num_m)
+            mi = np.arange(low_bound[ci], upp_bound[ci])[loc_start:loc_end]
+
+            # At last iteration of m's, it is necessary to assign m's to remaining ranks otherwise the MPI communication crashes.
+            if len(mi) == 0:
+                mi = np.array([low_bound[ci] + rank])
+
+            if len(mi) != 0:
+                print("Processing m-mode %i on rank %i" % (mi, rank))
+
+            nfreq = self.telescope.nfreq
+            lmax = self.telescope.lmax
+
+            if loc_num > 0:
+                if self.num_evals(mi) > 0:
+                    # Generate random KL data
+                    x = self.gen_sample(mi, n)
+                    vec1, vec2 = self.project_vector_kl_to_sky(mi, x)
+                    # Make array contiguous
+                    vec1 = np.ascontiguousarray(vec1.reshape(nfreq, lmax + 1, n))
+
+                # If I don't have evals - return zero vector
+                else:
+                    vec1 = np.zeros((nfreq, lmax + 1, n), dtype=np.complex128)
+
+            else:
+                vec1 = np.zeros((nfreq, lmax + 1, n), dtype=np.complex128)
+
+            et = time.time()
+
+            dsize = np.prod(vec1.shape)
+
+            for ir in range(size):
```

Comment:
Could use a comment here to describe what this loop is doing

Reply:
yes done.